### PR TITLE
fix(iOS): ComboBox Popup location when displayed on a Modal

### PR DESCRIPTION
**Pull request by ajpinedam**
GitHub Issue (If applicable): closes https://github.com/unoplatform/nventive-private/issues/449

## PR Type

What kind of change does this PR introduce?
- Bugfix

## What is the current behavior?

When loading a ComboBox on iOS from a Modal Page, the PopUp with the option does not respect the positioning of the ComboBox.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/Controls/ComboBox/ComboBox.cs [hunk 1]
```diff
@@ -779,6 +779,25 @@ Point getChildLocation()
...
+				// the Popup with the options.
+				if (comboRect.X < combo.Frame.X)
+				{
+					comboRect.X = combo.Frame.X;
+				}
+
+				if (comboRect.Y < combo.Frame.Y)
+				{
+					comboRect.Y = combo.Frame.Y;
+				}
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/ComboBox/ComboBox.cs:798]: Did you find a way to re-create the issue in a runtime test? I'm not sure if it will work in a simulator (we're running on tablet for this test).

File: src/Uno.UI/UI/Xaml/Controls/ComboBox/ComboBox.cs [hunk 2]
```diff
@@ -779,6 +779,25 @@ Point getChildLocation()
 				}
 
 				var comboRect = combo.GetAbsoluteBoundsRect();
+
+#if __IOS__
+				// We might get to a situation where the calculations performed during
+				// .GetAbsoluteBoundsRect() did not return the correct value
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/ComboBox/ComboBox.cs:785]: What about other callers of `GetAbsoluteBoundsRect`? Do they have the same issue? In this case should the fix go into `GetAbsoluteBoundsRect` itself?

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Controls/ComboBox/ComboBox.cs:785]: I think we need to understand the root cause of `GetAbsoluteBoundsRect` returning an incorrect value, in case the issue is deeper somewhere.

Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/Controls/ComboBox/ComboBox.cs:785]: It could be that GetAbsoluteBoundsRect is not able to determine how to get out of its own tree ancestors, and determine its own anchor.

Comment (ajpinedam) [reply src/Uno.UI/UI/Xaml/Controls/ComboBox/ComboBox.cs:785]: The reason I know it:

`GetAbsoluteBoundsRect ()` uses the `Windows.Current.Content `as the `EndPoint` when calculating the position of a Children View. This means it will start iterating from the View (in this case, the ComboBox) til it finds the `To` which will almost always be the Frame. 

The way we launch a Modal Page is by mimic-ing a Multi-Frame navigation.

A `ComboBox` hierarchy inside the Main Window looks like this

```xml
><Uno_UI_Controls_Window: 0x7f7f8d91cf50
  ><UITransitionView: 0x7f7f8cf0a820
    ><UIDropShadowView: 0x7f7f8cf0d0d0
      ><UILayoutContainerView: 0x7f7f8d9206d0
        ><UINavigationTransitionView: 0x7f7f8da781b0
        >Uno.UI.Xaml.Core.RootVisual-(393x852)@(0,0)
          >Windows.UI.Xaml.Controls.Border-(393x852)@(
            >Windows.UI.Xaml.Controls.Frame-(393x852)@(0,0)
              >Uno.Toolkit.UI.NativeFramePresenter-(393x852)@(0,0)
                ><UILayoutContainerView: 0x7f7f8cf229f0
                  ><UINavigationTransitionView: 0x7f7f8cf26ee0
                    ><UIViewControllerWrapperView: 0x7f7f8cf3c450
                      >SampleComboBoxFlyout.MainPage-(393x852)@(0,0)
                        >SampleComboBoxFlyout.MultiFrame-'NavigationRoot'-(393x852)@(0,0)
                          >Windows.UI.Xaml.Controls.Grid-(393x852)@(0,0)
                            >Windows.UI.Xaml.Controls.ComboBox-(123x32)
```

Where it will be part of the `Windows.UI.Xaml.Controls.Frame` 

When we put a `ComboBox` inside a Modal Page, the hierarchy looks like this now

```xml
><Uno_UI_Controls_Window: 0x7f7f8d91cf50
  ><UITransitionView: 0x7f7f8d818d50
    ><UIDimmingView: 0x7f7f8d822e70
    ><UIDropShadowView: 0x7f7f8d825550
      ><_UIRoundedRectShadowView: 0x7f7f8d8257b0
      ><UIView: 0x7f7f8d825a00
        ><UIView: 0x7f7f8d81fbb0
          *>SampleComboBoxFlyout.ModalPage-(393x783)@(0,0)
            >Windows.UI.Xaml.Controls.Grid-(393x783)@(0,0)
              >Windows.UI.Xaml.Controls.ComboBox-(0x0)@(0,0)
```

As can be seen, this hierarchy does not contains a `Frame` but we are telling that we want to iterate all the way up til we reach a `Frame`. This causes `GetAbsoluteBoundsRect ` to return invalid information for the Frame.

As the comment states, this workaround is a way to ensure that if we cannot get the right location from `GetAbsoluteBoundsRect ` we at least use the location information from the ComboBox itself.

Modifying the `ComboBox` sounds like the less intrusive solution we can have, taking into account that this is an edge case.

[Related method](https://github.com/unoplatform/uno/blob/master/src/Uno.UI/UI/Xaml/UIElement.iOS.cs#L164) found on UIElement.iOS.cs

Hope this helps clear any doubts.

cc: @jeromelaban  @Youssef1313

Comment (ajpinedam) [reply src/Uno.UI/UI/Xaml/Controls/ComboBox/ComboBox.cs:785]: Btw: Also tried using a `Frame` as part of the Modal Page where the ComboBox is displayed. Unfortunately, this breaks the whole app since we are not expecting to find more than one `Frame` in the tree hierarchy.

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ComboBox.cs
Comment (Xiaoy312) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ComboBox.cs:1]: you can drop the non-`__IOS__` parts, since the test is only for ios, about an ios-specific feature.

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ComboBox.cs [hunk 3]
```diff
@@ -718,4 +794,378 @@ protected override void OnCollectionChanged(NotifyCollectionChangedEventArgs e)
...
+
+			public void SetTransitionInfo(UIViewControllerSectionsTransitionInfo transitionInfo)
+			{
+				ModalInPresentation = !transitionInfo.AllowDismissFromGesture;
+				ModalPresentationStyle = transitionInfo.ModalPresentationStyle;
+				ModalTransitionStyle = transitionInfo.ModalTransitionStyle;
+			}
+		}
+
+		public abstract class FrameSectionsTransitionInfo : SectionsTransitionInfo
```
Comment (Xiaoy312) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ComboBox.cs:867]: you can probably trim down the fat a little, some of these stock transitions arent used at all, even more so if you removed no-ios code

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ComboBox.cs [hunk 4]
```diff
@@ -718,4 +794,378 @@ protected override void OnCollectionChanged(NotifyCollectionChangedEventArgs e)
...
+			public static Task FadeOutFrame1ToRevealFrame2(Frame frame1, Frame frame2)
+			{
+				// 1. Disable the currently visible frame during the animation.
+				frame1.IsHitTestVisible = false;
+
+				// 2. Make the next frame visible so that we see it as the previous frame fades out.
+				frame2.Opacity = 1;
+#if __IOS__ || __ANDROID__
+				// TODO: Fix this workaround
+				frame2.SetValue(UIElement.OpacityProperty, 1d, DependencyPropertyValuePrecedences.Animations);
```
Comment (Xiaoy312) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ComboBox.cs:1012]: you dont need this workaround anymore: #11865

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ComboBox.cs [hunk 5]
```diff
@@ -718,4 +794,378 @@ protected override void OnCollectionChanged(NotifyCollectionChangedEventArgs e)
 		}
 
 	}
+#if __IOS__
+	#region "Helper classes for the Modal Pop Up"
```
Comment (Xiaoy312) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ComboBox.cs:798]: they are just pages like any others; popup is not the best name, and it has special meaning in uwp...
```suggestion
	#region "Helper classes for the iOS Modal Page (UIModalPresentationStyle.pageSheet)"
```

File: src/Uno.UI/UI/Xaml/UIElement.cs [hunk 6]
```diff
@@ -1093,5 +1105,7 @@ public UI.Input.InputCursor ProtectedCursor
 			set;
 		}
 #endif
+
+		private partial class TransformToVisualContext { }
```
[the remark this answers is not in the document]
Comment (ajpinedam) [reply src/Uno.UI/UI/Xaml/UIElement.cs]: changed!

**Review by Xiaoy312 — APPROVED** [review]

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_Image.cs [hunk 7]
```diff
@@ -306,7 +306,7 @@ public async Task When_Image_Source_Nullify()
 			await WindowHelper.WaitForLoaded(parent);
 			var result = await TakeScreenshot(parent);
 
-			var sample = SUT.GetRelativeCoords(parent);
+			var sample = parent.GetRelativeCoords(SUT);
```
Comment (ajpinedam) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_Image.cs:309]: This Unit test used the `GetRelativeCoords()` method wrong, and I had t fix it.

File: src/Uno.UI/UI/Xaml/UIElement.cs [hunk 8]
```diff
@@ -534,10 +536,18 @@ internal static Matrix3x2 GetTransform(UIElement from, UIElement to)
...
-				var toToRoot = GetTransform(to, null);
-				var rootToTo = toToRoot.Inverse();
+				var toContext = new TransformToVisualContext();
+				var toToRoot = GetTransform(to, null, toContext);
 
-				matrix *= rootToTo;
+#if __IOS__
+				// On iOS, the `from` and `to` may be coming from different ViewController.
+				// In such case, their coordinates should not be "added" together, since they are from different coordinates space.
+				if (context.ViewController == toContext.ViewController)
```
Comment (ajpinedam) [inline src/Uno.UI/UI/Xaml/UIElement.cs:545]: @Xiaoy312 there are valid situations where the `ViewController` is null.

Comment (Xiaoy312) [reply src/Uno.UI/UI/Xaml/UIElement.cs:545]: i guess we can also the top-most crawled root in the context,
and compare if they are equal when both vc are null, to decide whether the offsets/matrices should be aggregated or not

Comment (ajpinedam) [reply src/Uno.UI/UI/Xaml/UIElement.cs:545]: You mean finding their respective roots and making sure they are part of the same tree? My only concern is figuring out how expensive this can get.

Comment (Xiaoy312) [reply src/Uno.UI/UI/Xaml/UIElement.cs:545]: we are already traversing the visual-tree, just return the last known element from TryGetParentUIElementForTransformToVisual before we return

**Review by jeromelaban — APPROVED** [review]

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/TestsResources.xaml [hunk 9]
```diff
@@ -622,4 +623,334 @@
...
+															 To="1">
+											<DoubleAnimation.EasingFunction>
+												<CubicEase EasingMode="EaseIn" />
+											</DoubleAnimation.EasingFunction>
+										</DoubleAnimation>
+									</Storyboard>
+								</VisualState>
+
+								<VisualState x:Name="Pressed">
+									<!-- Workaround for Issue #246732 where IOS combobox sometimes keeps the pressed-on state too long. -->
```
Comment (agneszitte) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/TestsResources.xaml:677]: It would be best to add the Issue link directly in the comment no? (instead of the number to go to the correct repo)

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/TestsResources.xaml [hunk 10]
```diff
@@ -622,4 +623,334 @@
...
+											  Grid.Column="1"
+											  VerticalAlignment="Center"
+											  HorizontalAlignment="Right" />
+
+						<!-- Popup -->
+						<Popup x:Name="Popup">
+							<Border x:Name="PopupBorder"
+										CornerRadius="{StaticResource ComboBoxRadius}"
+										Background="{StaticResource MaterialSurfaceBrush}">
+								<!-- Workaround for Issue #213 where android PopupBorder doesn't get redrawn properly, adds Background, CornerRadius and Border to ScrollViewer  -->
```
Comment (agneszitte) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/TestsResources.xaml:928]: Same here, would be best to have the link to the issue (instead of just the number)

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ComboBox.cs (hunk 3 above)
Comment (ajpinedam) [reply src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ComboBox.cs:860]: did that thanks

File: src/Uno.UI/UI/Xaml/UIElement.cs (hunk 8 above)
Comment (ajpinedam) [reply src/Uno.UI/UI/Xaml/UIElement.cs:545]: As we discuss on Teams, we will check this as an edge case and it will be worked under a different GH issue https://github.com/unoplatform/uno/issues/12250

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/TestsResources.xaml (hunk 9 above)
Comment (ajpinedam) [reply src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/TestsResources.xaml:677]: @agneszitte Thanks for the comment. This was a code I copied from a Sample App. Let me check if the issue this is still an issue, and if so, would you allow me to push a new PR to update this? That way, we could merge this PR that has been idle for a few weeks.

Comment (agneszitte) [reply src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/TestsResources.xaml:677]: Yes that works for me @ajpinedam :) !
